### PR TITLE
🎉 Destination Redshift: Use SUPER data type on Redshift destination for raw JSON data

### LINKED ISSUE #5050: Use SUPER data type on Redshift destination for raw JSON data
## Tell us about the problem you're trying to solve
I'd like the option utilize the SUPER column type when loading data to a Redshift destination. https://docs.aws.amazon.com/redshift/latest/dg/r_SUPER_type.html  

## Describe the solution you’d like
Allow for checkbox selection on either the destination or connection to utilize SUPER data type as the target for the column _airbyte_data. Experimenting with this data type I've found performance to be increased during normalization, as well for allowing greater flexibility when creating source queries/views without normalization

## Describe the alternative you’ve considered or used
N/A

## Additional context
N/A

## Are you willing to submit a PR?
Just recently got a local build setup, will explroe

### PR DESCRIPTION
**Pull request by alexandertsukanov**
## What
Added an option to use SUPER Data type in _airbyte_data column, in _airbyte_raw table, instead of VARCHAR.  This option should increase performance during normalization, as well allows greater flexibility when creating source queries/views without normalization.

This **PR** contains only **Java part**. The ticket for **Python team** will be created and linked to this PR.

## How
Changed connector specification to allow user set the SUPER Data type for Destination Redshift connector. Changed SQL query for each flow.

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>

#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/9407)

### CODE REVIEW

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftInsertDestination.java [hunk 1]
```diff
@@ -11,20 +11,21 @@
...
 import io.airbyte.integrations.base.Destination;
 import io.airbyte.integrations.destination.jdbc.AbstractJdbcDestination;
+import io.airbyte.integrations.destination.redshift.enums.RedshiftDataTmpTableMode;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-public class RedshiftInsertDestination extends AbstractJdbcDestination implements Destination {
```
Comment (alexandertsukanov) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftInsertDestination.java:20]: Removed as AbstractJdbcDestination implement Destination. So no sense to implement in child class.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/enums/RedshiftDataTmpTableMode.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    @Override
+    public String getTmpTableSqlStatement(String schemaName, String tableName) {
+      return String.format(
+          "CREATE TABLE IF NOT EXISTS %s.%s ( \n"
+              + "%s VARCHAR PRIMARY KEY,\n"
+              + "%s SUPER,\n"
+              + "%s TIMESTAMP WITH TIME ZONE DEFAULT CURRENT_TIMESTAMP\n"
+              + ");\n",
+          schemaName, tableName, JavaBaseConstants.COLUMN_NAME_AB_ID, JavaBaseConstants.COLUMN_NAME_DATA, JavaBaseConstants.COLUMN_NAME_EMITTED_AT);
+    }
```
Comment (sashaNeshcheret) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/enums/RedshiftDataTmpTableMode.java:23]: Seems to avoid duplication we can move the same logic in parent method, and move SUPER or  VARCHAR(max) in overridden methods.

**Review by sashaNeshcheret — APPROVED** [review]

**Review by alexandr-shegeda — APPROVED** [review]
LGTM

**Comment by alexandr-shegeda** [comment]
@alexandertsukanov we can proceed with publishing and merging without airbyte review

**Comment by sherifnada** [comment]
@alexandertsukanov 

Like described in the [UX handbook](https://docs.airbyte.com/connector-development/ux-handbook#input-configuration) it's preferable if we make a decision on behalf of the user instead of ask them which data type to use. If there is a good reason to give them an option instead of making a default choice, then exposing this as an option is acceptable. But we first need to make this research. 

Can you do some investigation to understand if there is  any situation in which the user would _not_ want to use SUPER over JSON?

**Comment by alexandertsukanov** [comment]
> @alexandertsukanov
> 
> Like described in the [UX handbook](https://docs.airbyte.com/connector-development/ux-handbook#input-configuration) it's preferable if we make a decision on behalf of the user instead of ask them which data type to use. If there is a good reason to give them an option instead of making a default choice, then exposing this as an option is acceptable. But we first need to make this research.
> 
> Can you do some investigation to understand if there is any situation in which the user would _not_ want to use SUPER over JSON?

Hi, @sherifnada ! Yep, I will make a research to understand if we may have such situations. Thanks.

**Comment by alexandertsukanov** [comment]
@sherifnada please see my comment here https://github.com/airbytehq/airbyte/pull/9610#issuecomment-1027789721. Thanks.

**Comment by ChristopheDuong** [comment]
As commented https://github.com/airbytehq/airbyte/pull/9610#issuecomment-1028923806, it sounds like the destination-redshift should only allow `SUPER` data types and make sure to upgrade any existing raw tables to that new schema then

**Comment by sherifnada** [comment]
@alexandertsukanov i defer to @ChristopheDuong on this -- happy to move forward with `SUPER` here based on the research

**Comment by alexandertsukanov** [comment]
@ChristopheDuong please review the changes. Was added the integration test to check workflow you have described here: https://github.com/airbytehq/airbyte/pull/9610 
Thanks.

File: airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftDestinationTest.java [hunk 3]
```diff
@@ -18,22 +20,23 @@
...
-  @DisplayName("When given S3 credentials should use COPY")
-  public void useCopyStrategyTest() {
+  @DisplayName("When given S3 credentials should use COPY with SUPER Datatype")
+  public void useCopyStrategyTestWithSuperTmpTable() {
     final var stubConfig = mapper.createObjectNode();
     stubConfig.put("s3_bucket_name", "fake-bucket");
     stubConfig.put("s3_bucket_region", "fake-region");
     stubConfig.put("access_key_id", "test");
     stubConfig.put("secret_access_key", "test key");
+    stubConfig.put("use_super_redshift_type", "true");
```
Comment (grubberr) [inline airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftDestinationTest.java:30]: why do we have it here ? as I understand we removed  this option "use_super_redshift_type" ?

File: airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftDestinationTest.java [hunk 4]
```diff
@@ -18,22 +20,23 @@
...
   }
 
   @Test
-  @DisplayName("When not given S3 credentials should use INSERT")
-  public void useInsertStrategyTest() {
+  @DisplayName("When not given S3 credentials should use INSERT with SUPER Datatype")
+  public void useInsertStrategyTestWithSuperDatatype() {
     final var stubConfig = mapper.createObjectNode();
-    assertFalse(RedshiftDestination.isCopy(stubConfig));
+    stubConfig.put("use_super_redshift_type", "true");
```
Comment (grubberr) [inline airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftDestinationTest.java:39]: why do we have it here ? as I understand we removed  this option "use_super_redshift_type" ?

File: airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftDestinationTest.java (hunk 3 above)
Comment (alexandertsukanov) [reply airbyte-integrations/connectors/destination-redshift/src/test/java/io/airbyte/integrations/destination/redshift/RedshiftDestinationTest.java]: Thanks, removed

**Review by andriikorotkov — APPROVED** [review]

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java [hunk 5]
```diff
@@ -4,41 +4,67 @@
...
   private static final Logger LOGGER = LoggerFactory.getLogger(RedshiftSqlOperations.class);
   protected static final int REDSHIFT_VARCHAR_MAX_BYTE_SIZE = 65535;
+  private static final String SELECT_ALL_TABLES_WITH_NOT_SUPER_TYPE_SQL_STATEMENT = """
+      select tablename, schemaname
+      from pg_table_def
+      where
+      schemaname = '%s'
+      and "column" = '%s'
+      and type <> 'super'
+      and tablename like '%%raw%%'""";
```
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java:40]: Is grepping just "raw" sufficient? it is really generic word to look for... Not just airbyte is using that in the naming of table.

maybe looking for `%%airbyte_raw%%` would be more specific? and add some conditions on the column that we typically use? (`ab_id`, `airbyte_data`, etc)

I am guessing it would be pretty bad if you change datatypes from a user schema that is not related to airbyte...

Comment (alexandertsukanov) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftSqlOperations.java]: Done. Thanks.

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/SqlOperations.java [hunk 6]
```diff
@@ -117,4 +117,18 @@ default boolean isSchemaExists(final JdbcDatabase database, final String schemaN
    */
   boolean isSchemaRequired();
 
+
+  /**
+   * The method is responsible for executing some specific DB Engine logic in onStart method.
+   * We can to override this method to execute specific logic e.g. to handle any necessary migrations in the destination, etc.
```
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/SqlOperations.java:123]:
```suggestion
   * We can override this method to execute specific logic e.g. to handle any necessary migrations in the destination, etc.
```

Comment (alexandertsukanov) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/SqlOperations.java]: Fixed. Thanks.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftDestination.java [hunk 7]
```diff
@@ -63,13 +60,13 @@ public static boolean isCopy(final JsonNode config) {
     final var emptyRegion = regionNode == null || regionNode.asText().equals("");
 
     if (bucketNode == null && emptyRegion && accessKeyIdNode == null && secretAccessKeyNode == null) {
-      return false;
+      return DestinationType.INSERT_WITH_SUPER_TMP_TYPE;
```
Comment (ChristopheDuong) [inline airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftDestination.java:63]: I made a similar comment on the BigQuery destination, we should probably introduce a similar warning here to recommend using staging instead of direct insert too!
see https://github.com/airbytehq/airbyte/pull/11776/files#r844957299

**Review by ChristopheDuong — APPROVED** [review]

File: airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/SqlOperations.java (hunk 6 above)
Comment (alexandertsukanov) [reply airbyte-integrations/connectors/destination-jdbc/src/main/java/io/airbyte/integrations/destination/jdbc/SqlOperations.java]: Fixed. Thanks.

File: airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftDestination.java (hunk 7 above)
Comment (alexandertsukanov) [reply airbyte-integrations/connectors/destination-redshift/src/main/java/io/airbyte/integrations/destination/redshift/RedshiftDestination.java:65]: WARN was added. Thanks for recommendation.

**Comment by sherifnada** [comment]
@alexandertsukanov i won't be able to review this PR, please proceed without my review if the team approves
